### DISCUSSION TITLE
Custom Camera Feed

### DISCUSSION THREAD

**Discussion by mysticprysm** [OrcaSlicer/OrcaSlicer#2548] [Ideas]
I can't imagine it would be too difficult to add the function to input a custom stream address into the video player section? Would be great for those of us wanting to monitor our prints with our custom cameras. And then maybe later adding functionality to do basic in app timelapses of sorts. I'd absolutely be willing to learn if anyone would be willing to walk through the code with me.

**Comment by stuomas** [comment]
I also wanted this for my P1S because the camera sucks, so I took a shot at it.

It started simple with hardcoded custom URL, and then it got hard when I needed to start thinking about the UX stuff :)

Anyways here is my fork:

~~https://github.com/stuomas/OrcaSlicer/tree/custom-camera-source~~
https://github.com/stuomas/OrcaSlicer/tree/custom-camera-source-webview

and it Works On My Machine™ pretty nicely. Tested only on Windows 11, for some reason cannot log in with the Linux AppImage and Mac I don't have at hand.

I noticed some small hiccups sometimes when starting/stopping bambulab's camera manually (why does there even need to be manual start/stop?) and it stays black but custom camera works well. Bambulab's way of handling the camera stream is so weird that I didn't want to start deciphering how it's supposed to work, so I just mostly worked around it. In the end both cameras work and you can switch between them.

I use Raspberry Pi camera with [camera-streamer](https://github.com/ayufan/camera-streamer) and the `http://<ip>:<port>/video.mp4` at least works. If you want to test this out, you can use any URL to an mp4 file for example from [here](https://gist.github.com/jsturgis/3b19447b304616f18657)

![orca_screencast](https://github.com/SoftFever/OrcaSlicer/assets/16635041/4fd1089d-4a7d-4421-b80e-5489f960a0e3)

You can find build artifacts here
~~https://github.com/stuomas/OrcaSlicer/actions/runs/7197604520~~
~~https://github.com/stuomas/OrcaSlicer/actions/runs/7231707473~~

If there is any interest having this kind of feature upstream, I might be able to continue work on it and make a PR. Would need help at least with deciding how the UX/UI should look. Also the camera URL isn't persisted anywhere yet so need to input it every time the slicer starts, might look into that next.

**Comment by mysticprysm** [comment]
That's awesome! I'd love to contribute anyway possible! Unfortunately I
don't have coding experience lol. But I can help with any sort of testing
etc.

**Reply by stuomas** [reply]
Would be very helpful if you test it out and report experiences!

**Comment by mysticprysm** [comment]
Excuse my ignorance, but with your fork, how would I get it? Do I have to
compile myself or do I download it as a zip?

**Reply by stuomas** [reply]
I put a link to the build artifacts in the end. For example for Windows, download the [OrcaSlicer_Windows_V1.9.0-dev](https://github.com/stuomas/OrcaSlicer/suites/19014937250/artifacts/1112469736) and unzip and install. I haven't tested those artifacts myself tbh but it's the same forked build workflow as SoftFever uses so I assume they work heh

**Comment by mysticprysm** [comment]
Appreciate it! I don't have any real coding experience and my experience on
github has been limited to downloading pre existing stuff lol.

**Comment by mysticprysm** [comment]
Absolutely! I don't even have the camera URL input and I think
its beautiful! lol! Been dying to get a custom feed in there! Because I
want to use like an old phone or a pi as an IP camera

**Comment by mysticprysm** [comment]
Having a hair bit of trouble getting a URL to work. What kind of URL should
I use? RTSP? Another kind? I'm using the ip webcam app on android and none
of those urls seem to work.

**Reply by stuomas** [reply]
I'm not sure of the capabilities of wxMediaCtrl or the backend it is using. I tried MJPEG and RTSP streams and they did not work. Maybe it needs to be a h264 stream on HTTP server. I can look into it more. Single frames from IP Webcam work (like `http://192.168.1.5:8080/shot.jpg`) so the connection works. RTSP might need some middleware to work, looks like what bambulabs has done with their closed source component.

**Reply by stuomas** [reply]
I think I found an ok solution to make also MJPEG work (IP Webcam app's `/video` endpoint for example) that also greatly simplifies many other things: using WebView instead of wxMediaCtrl. I'll work on that on the weekend so stay tuned hehe

**Comment by mysticprysm** [comment]
Niiiice! I'm getting email updates from this thread so I'll know when you
or anyone else posts here and I'll test accordingly

**Comment by stuomas** [comment]
New iteration on new branch https://github.com/stuomas/OrcaSlicer/tree/custom-camera-source-webview

Supports also MJPEG stream from android's IP Webcam app (`http://<ip>:<port>/video`)

Also the URL and enable/disable status is persisted and restored when program starts.

Still small bug that upon starting the program with persisted URL, the javascript that removes the video controls/scrollbars does not work, only works when changing source manually🤔

**Reply by stuomas** [reply]
@mysticprysm I think I fixed issue with controls/scrollbars, new build is in progress here: https://github.com/stuomas/OrcaSlicer/actions/runs/7292485309

I will continue after vacations, merry Christmas!

**Comment by stuomas** [comment]
Opened PR https://github.com/SoftFever/OrcaSlicer/pull/3415

**Comment by profucius** [comment]
@stuomas Hi there, I found this thread while trying to setup an IP cam through Orcaslicer. I found the Custom Camera feature in the app and I think it is because of you that we have this feature, if so, thank you!

Sorry to resurrect this from a year ago. I'm curious if you're still working on it or plan to update it? I'm having a bit of trouble getting it setup. I am using go2rtc to take an `rtsp` feed and convert it to an `http` feed, but I have like 20 different options with all sorts of permutations, and haven't yet found the right one that the app expects.

Which specific type of feed is required? I am assuming `rtsp` never got working, so `http` is probably the first expectation. Below is a screenshot of all the options I have, if you are able to please advise. Thanks!

Most important question: How can I debug the camera feed, is there a place in a log somewhere that tells me why the URL I am testing is failing?

---

Edit: Should I be using VLC to test the URL as a "network stream"? Aka, if VLC can stream it, then Orcaslicer can, and vis versa if it cant then neither can it?

Edit 2: If we could get RTSP working directly, that would be so excellent! Perhaps there is some logic in the way that VLC processes RTSP links that could be gleaned and used to make it work in Orcaslicer?

---

![image](https://github.com/user-attachments/assets/4180e49a-4495-4920-ba0c-b68c165ead05)

**Reply by stuomas** [reply]
Hi, thanks for testing it out, the feature haven't really gotten any feedback yet. MJPEG stream should work as well as MP4. The earlier is more realtime and latter has some delay but can be used in picture-in-picture mode which is nice.

If the link opens natively in browser, it likely works (because in orca it's played in webview) 

I have no plans to working on the feature at the moment, except in case of some serious issues

**Reply by profucius** [reply]
No worries. I've pivoted to making a dashboard for my Bambu printer and Eufy camera within Home Assistant, which gives me many more stats and other data all in one place. Thanks though!

**Comment by hexsmith** [comment]
Hey, guys.  I wish the "Custom Camera" could be activated even when the printer does not appear on the network.  I have a P1S in LAN mode on a Wi-Fi island and an RTSP camera on a network with Internet access.  I flip back and forth between networks, and I use OBS Studio or VLC to tune into the RTSP camera, but it would be super cool if I could just tune into it in the slicer even when I'm not on the LAN with the printer.
